File: 백준/Silver/14465. 소가 길을 건너간 이유 5/소가 길을 건너간 이유 5.py
import sys
input = sys.stdin.readline
n,k,b = map(int,input().split())
start = 1
end = 1
arr = [1] * (n + 1)
for _ in range(b):
    arr[int(input())] = 0
#print(arr)
ans = 100001
cnt = 0
cnt_b = 0 #부서진 것 개수
while end <= n:
    tmp = end - start + 1
    if arr[end] == 0:
        cnt += 1

    if  tmp == k:
        #print(start, end, cnt)
        ans = min(ans, cnt)
        if arr[start] == 0:
            cnt -= 1
        start += 1
        end += 1

    elif tmp < k:
        end += 1

    else: # cnt > k
        if arr[start] == 0:
            cnt -= 1
        start += 1
print(ans)

File: 백준/Silver/14465. 소가 길을 건너간 이유 5/소가 길을 건너간 이유 5.java
import java.io.*;
import java.util.*;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        int n = Integer.parseInt(st.nextToken());
        int k = Integer.parseInt(st.nextToken());
        int b = Integer.parseInt(st.nextToken());
        int ans = 0;

        boolean[] check = new boolean[n + 1];
        Arrays.fill(check, true);
        for (int i = 0; i < b; i++) {
            check[Integer.parseInt(br.readLine())] = false;
        }

        for (int i = 1; i < 1 + k; i++) {
            if(!check[i]) ans++;
        }
        int cnt = ans;

        int start = 1;
        int end = k;
        while(end < n){
            if(!check[start++]) cnt--;
            if(!check[++end]) cnt++;
            ans = Math.min(ans, cnt);
        }
        System.out.println(ans);
    }
}
